Add contribution guide for third-party library and fix pre-commit (#663)
pytorch/data: third_party/CONTRIBUTING.md

# Third-Party Libraries

> The `third_party` directory contains all of (optional) dependency libraries used by `torchdata`.

Reviewer: ```suggestion
The `third_party` directory contains all of the (optional) dependency libraries used by `torchdata`.
```

And, it relies on the `CMake` system to compile and integrate them into a C-extension module that
can be found as `torchdata/_torchdata`.

`torchdata` also relies on [`pybind11`](https://github.com/pybind/pybind11) to expose C++ API in
Python. Please refer this [directory](https://github.com/pytorch/data/tree/main/torchdata/csrc)
for more detail.

## Integration Requirement

### Soft Dependency vs Hard Dependency

`TorchData` as a data-processing libraries provides a bunch of `DataPipes` that are integrated with
different third-party libraries to handle specific use cases. For example, `datasets` is imported
to load dataset from `HuggingFace`, and `fsspec` is imported to provide a unified API to access and

> load from local or remote file systems. Those dependencies should only be verified of their availability
> until they are used/referenced by users during `DataPipe` construction. You can find examples from

Reviewer: ```suggestion
load from local or remote file systems. Those dependencies are optional and should only be verified of their availability
when they are used/referenced by users during `DataPipe` construction. You can find examples from
```

[here](https://github.com/pytorch/data/blob/bb78231e5f87620385cb2f91cda87e7f9414eb4a/torchdata/datapipes/iter/load/huggingface.py#L57-L62)

> and [here](https://github.com/pytorch/data/blob/d19858202df7e8b75765074259e6023f539cbf3f/torchdata/datapipes/iter/load/fsspec.py#L59).

Reviewer: ```suggestion
and [here](https://github.com/pytorch/data/blob/d19858202df7e8b75765074259e6023f539cbf3f/torchdata/datapipes/iter/load/fsspec.py#L59).
They are contrasted with Core dependencies that are must be installed along with `torchdata`.
```

# Third-Party Libraries

The `third_party` directory contains all of the (optional) dependency libraries used by `torchdata`.
And, it relies on the `CMake` system to compile and integrate them into a C-extension module that
can be found as `torchdata/_torchdata`.

`torchdata` also relies on [`pybind11`](https://github.com/pybind/pybind11) to expose C++ API in
Python. Please refer this [directory](https://github.com/pytorch/data/tree/main/torchdata/csrc)
for more detail.

## Integration Requirement

### Soft Dependency vs Hard Dependency

`TorchData` as a data-processing libraries provides a bunch of `DataPipes` that are integrated with
different third-party libraries to handle specific use cases. For example, `datasets` is imported
to load dataset from `HuggingFace`, and `fsspec` is imported to provide a unified API to access and
load from local or remote file systems. Those dependencies are optional and should only be verified

> of their availability when they are used/referenced by users during `DataPipe` construction.You can

Reviewer: ```suggestion
of their availability when they are used/referenced by users during `DataPipe` construction. You can
```
